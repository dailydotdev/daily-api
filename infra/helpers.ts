import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
