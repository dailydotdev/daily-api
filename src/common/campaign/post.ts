import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
