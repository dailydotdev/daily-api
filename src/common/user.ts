import { DataSource } from 'typeorm';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';
import { ghostUser } from './index';
import { cancelSubscription } from './paddle';
import type { AuthContext, Context } from '../Context';
import { ForbiddenError } from 'apollo-server-errors';
import { logger } from '../logger';
import { CoresRole } from '../types';
import { remoteConfig } from '../remoteConfig';
import { UserTransaction } from '../entity/user/UserTransaction';
import { SubscriptionProvider, SubscriptionStatus } from './plus';
import { deleteUserResume } from './googleCloud';

export const deleteUser = async (
  con: DataSource,
  userId: string,
  messageId?: string,
) => {
  try {
    const { subscriptionFlags } = await con.getRepository(User).findOneOrFail({
      select: ['subscriptionFlags'],
      where: { id: userId },
    });

    // If the user has a subscription, we need to handle it before deletion
    if (subscriptionFlags?.subscriptionId) {
      if (
        subscriptionFlags?.provider === SubscriptionProvider.AppleStoreKit &&
        subscriptionFlags?.status === SubscriptionStatus.Active
      ) {
        throw new Error(
          'Apple subscriptions are not supported for user deletion',
        );
      }

      if (subscriptionFlags?.provider === SubscriptionProvider.Paddle) {
        const isGifted = !!subscriptionFlags.giftExpirationDate;
        // gifted subscription is a one-time payment hence not considered subscription in Paddle's terms
        if (!isGifted) {
          await cancelSubscription({
            subscriptionId: subscriptionFlags.subscriptionId,
          });
        }
        logger.info(
          {
            provider: SubscriptionProvider.Paddle,
            userId,
            isGifted,
            subscriptionId: subscriptionFlags.subscriptionId,
          },
          'Subscription cancelled user deletion',
        );
      }
    }

    // Delete user's resume if exists
    await deleteUserResume(userId);

    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: ghostUser.id,
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: ghostUser.id });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(UserTransaction).update(
        {
          senderId: userId,
        },
        {
          senderId: ghostUser.id,
        },
      );
      await entityManager.getRepository(UserTransaction).update(
        {
          receiverId: userId,
        },
        {
          receiverId: ghostUser.id,
        },
      );
      await entityManager.getRepository(User).delete(userId);
    });
    logger.info(
      {
        userId,
        messageId,
      },
      'deleted user',
    );
  } catch (err) {
    logger.error(
      {
        userId,
        messageId,
        err,
      },
      'failed to delete user',
    );
    throw err;
  }
};

export /**
 * Function creator that wraps a function with auth protection
 * In this case it checks ctx contains userId but it can be expanded
 * in the future to check more advanced things.
 * This is a DX layer to make sure auth required stuff like cores
 * is protected on the function level.
 *
 * @template Props
 * @template Result
 * @param {(props: Props) => Result} fn
 * @return {Result}
 */
const createAuthProtectedFn = <
  Props extends { ctx: Pick<AuthContext, 'userId'> },
  Result,
>(
  fn: (props: Props) => Result,
) => {
  return (props: Props): Result => {
    if (!props.ctx.userId) {
      throw new ForbiddenError('Auth is required');
    }

    return fn(props);
  };
};

export const getUserCoresRole = ({ region }: { region?: string }) => {
  if (!region) {
    return CoresRole.None;
  }

  const ruleForRegion = remoteConfig.vars.coresRoleRules?.find((rule) =>
    rule.regions.includes(region),
  );

  if (ruleForRegion) {
    return ruleForRegion.role;
  }

  return CoresRole.Creator;
};

export const checkUserCoresAccess = ({
  user,
  requiredRole,
}: {
  user: Pick<User, 'id' | 'coresRole'>;
  requiredRole: CoresRole;
}): boolean => {
  if (!user) {
    return false;
  }

  if (typeof user.coresRole !== 'number') {
    return false;
  }

  if (user.coresRole < requiredRole) {
    return false;
  }

  return true;
};

export const checkCoresAccess = async ({
  ctx,
  userId,
  requiredRole,
}: {
  ctx: Context;
  userId: string;
  requiredRole: CoresRole;
}): Promise<boolean> => {
  const user = await ctx.dataLoader.user.load({
    userId,
    select: ['id', 'coresRole'],
  });

  if (!user) {
    return false;
  }

  return checkUserCoresAccess({ user, requiredRole });
};
