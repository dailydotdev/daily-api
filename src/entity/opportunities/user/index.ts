export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
