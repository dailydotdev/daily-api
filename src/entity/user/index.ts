export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
