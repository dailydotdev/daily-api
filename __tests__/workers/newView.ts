import worker from '../../src/workers/newView';
import { expectSuccessfulBackground, saveFixtures } from '../helpers';
import { postsFixture } from '../fixture/post';
import {
  Alerts,
  ArticlePost,
  Source,
  User,
  UserStreak,
  View,
} from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { usersFixture } from '../fixture/user';
import { DataSource, IsNull, Not } from 'typeorm';
import createOrGetConnection from '../../src/db';

let con: DataSource;

beforeAll(async () => {
  con = await createOrGetConnection();
});

beforeEach(async () => {
  await con.getRepository(UserStreak).clear();
  await con.getRepository(View).clear();
  await saveFixtures(
    con,
    User,
    usersFixture.map((u) => ({ ...u, id: `u${u.id}` })),
  );
  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, ArticlePost, postsFixture);
});

it('should save a new view without timestamp', async () => {
  await expectSuccessfulBackground(worker, {
    postId: 'p1',
    userId: 'u1',
    referer: 'referer',
    agent: 'agent',
    ip: '127.0.0.1',
  });
  const views = await con.getRepository(View).find();
  expect(views.length).toEqual(1);
  expect(views[0]).toMatchSnapshot({
    timestamp: expect.any(Date),
  });
});

it('should save a new view with the provided timestamp', async () => {
  const timestamp = new Date(2020, 5, 11, 1, 17);
  await expectSuccessfulBackground(worker, {
    postId: 'p1',
    userId: 'u1',
    referer: 'referer',
    agent: 'agent',
    ip: '127.0.0.1',
    timestamp: timestamp.toISOString(),
  });
  const views = await con.getRepository(View).find();
  expect(views.length).toEqual(1);
  expect(views[0]).toMatchSnapshot();

  const streak = await con
    .getRepository(UserStreak)
    .findOne({ where: { userId: 'u1', lastViewAt: timestamp } });
  expect(streak).toMatchSnapshot({
    updatedAt: expect.any(Date),
  });
});

it('should not save a new view within a week since the last view', async () => {
  const date1 = new Date(2020, 5, 11, 1, 17);
  const date2 = new Date(2020, 5, 13, 1, 17);
  const data = {
    postId: 'p1',
    userId: 'u1',
    referer: 'referer',
    agent: 'agent',
    ip: '127.0.0.1',
    timestamp: date1.toISOString(),
  };
  await expectSuccessfulBackground(worker, data);
  await expectSuccessfulBackground(worker, {
    ...data,
    timestamp: date2.toISOString(),
  });
  const views = await con.getRepository(View).find();
  expect(views.length).toEqual(1);
  expect(views[0]).toMatchSnapshot();
});

it('should save a new view after a week since the last view', async () => {
  const date1 = new Date(2020, 5, 11, 1, 17);
  const date2 = new Date(2020, 5, 19, 1, 17);
  const data = {
    postId: 'p1',
    userId: 'u1',
    referer: 'referer',
    agent: 'agent',
    ip: '127.0.0.1',
    timestamp: date1.toISOString(),
  };
  await expectSuccessfulBackground(worker, data);

  const streak1 = await con
    .getRepository(UserStreak)
    .findOne({ where: { userId: 'u1', lastViewAt: date1 } });
  expect(streak1).not.toBeNull();
  expect(streak1?.currentStreak).toEqual(1);

  await expectSuccessfulBackground(worker, {
    ...data,
    timestamp: date2.toISOString(),
  });

  const views = await con.getRepository(View).find();

  expect(views.length).toEqual(2);
  expect(views[1]).toMatchSnapshot();
});

describe('reading streaks', () => {
  const defaultStreak: Partial<UserStreak> = {
    currentStreak: 4,
    totalStreak: 42,
    maxStreak: 10,
  };

  const prepareTest = async (
    currentDate: Date | string | undefined,
    previousDate: Date | string | undefined,
    previousStreak: Partial<UserStreak> | null | undefined = defaultStreak,
  ) => {
    if (previousStreak) {
      await con.getRepository(UserStreak).save({
        ...previousStreak,
        userId: previousStreak.userId ?? 'u1',
        lastViewAt: previousDate ? new Date(previousDate) : undefined,
      });
    }

    const data = {
      postId: 'p1',
      userId: 'u1',
      referer: 'referer',
      agent: 'agent',
      ip: '127.0.0.1',
      timestamp: currentDate ? new Date(currentDate) : undefined,
    };
    await expectSuccessfulBackground(worker, data);
  };

  const runTest = async (
    currentDate: Date | string,
    previousDate: Date | string | undefined,
    previousStreak: Partial<UserStreak> | null | undefined = defaultStreak,
    expectedStreak?: Partial<UserStreak>,
  ) => {
    await prepareTest(currentDate, previousDate, previousStreak);

    const streak = await con.getRepository(UserStreak).findOne({
      where: {
        userId: 'u1',
        lastViewAt: new Date(currentDate),
      },
    });
    for (const key in expectedStreak) {
      expect(streak?.[key]).toEqual(expectedStreak[key]);
    }
  };

  it('updates reading streak without a timestamp', async () => {
    await prepareTest(undefined, undefined);

    const streak = await con
      .getRepository(UserStreak)
      .findOne({ where: { userId: 'u1', lastViewAt: Not(IsNull()) } });
    expect(streak).toMatchSnapshot({
      updatedAt: expect.any(Date),
      lastViewAt: expect.any(Date),
    });
  });

  it('does not update reading streak if view was not written', async () => {
    await prepareTest('2024-01-25T17:17Z', '2024-01-24T14:17Z');

    const streak1 = await con
      .getRepository(UserStreak)
      .findOne({ where: { userId: 'u1', currentStreak: 5 } });
    expect(streak1).not.toBeNull();

    const data = {
      postId: 'p1',
      userId: 'u1',
      referer: 'referer',
      agent: 'agent',
      ip: '127.0.0.1',
      timestamp: new Date('2024-01-26T17:17Z'),
    };
    await expectSuccessfulBackground(worker, data);

    const streak2 = await con
      .getRepository(UserStreak)
      .findOne({ where: { userId: 'u1' } });
    expect(streak2?.updatedAt).toEqual(streak1?.updatedAt);
    expect(streak2?.currentStreak).toEqual(streak1?.currentStreak);
  });

  it('does not update reading streak if view does not have userId', async () => {
    const data = {
      postId: 'p1',
      referer: 'referer',
      agent: 'agent',
      ip: '127.0.0.1',
      timestamp: new Date('2024-01-26T17:17Z'),
    };
    await expectSuccessfulBackground(worker, data);

    const streak = await con.getRepository(UserStreak).findOne({
      where: { lastViewAt: new Date('2024-01-26T17:17Z') },
    });
    expect(streak).toBeNull();
  });

  it('does not update reading streak if userId does not match existing user', async () => {
    const data = {
      postId: 'p1',
      userId: '__this_userId_should_not_exist__',
      referer: 'referer',
      agent: 'agent',
      ip: '127.0.0.1',
      timestamp: new Date('2024-01-26T17:17Z'),
    };
    await expectSuccessfulBackground(worker, data);

    const streak = await con.getRepository(UserStreak).findOne({
      where: { lastViewAt: new Date('2024-01-26T17:17Z') },
    });
    expect(streak).toBeNull();
  });

  it('should start a reading streak if there was no row in user_streak table', async () => {
    await runTest('2024-01-26T17:17Z', undefined, null, {
      currentStreak: 1,
      totalStreak: 1,
      maxStreak: 1,
      lastViewAt: new Date('2024-01-26T17:17Z'),
    });
  });

  it('should start a reading streak if there was none before', async () => {
    await runTest(
      '2024-01-26T17:17Z',
      undefined,
      {
        currentStreak: 0,
        totalStreak: 0,
        maxStreak: 0,
      },
      {
        currentStreak: 1,
        totalStreak: 1,
        maxStreak: 1,
        lastViewAt: new Date('2024-01-26T17:17Z'),
      },
    );
  });

  it('should increment a reading streak if lastViewAt was yesterday', async () => {
    await runTest('2024-01-26T19:17Z', '2024-01-25T17:23Z', defaultStreak, {
      currentStreak: 5,
      totalStreak: 43,
      maxStreak: 10,
      lastViewAt: new Date('2024-01-26T19:17Z'),
    });
  });

  it('should increment maxStreak if lastViewAt was yesterday and current streak is bigger', async () => {
    await runTest(
      '2024-01-26T19:17Z',
      '2024-01-25T17:23Z',
      {
        currentStreak: 4,
        totalStreak: 98,
        maxStreak: 4,
      },
      {
        currentStreak: 5,
        totalStreak: 99,
        maxStreak: 5,
        lastViewAt: new Date('2024-01-26T19:17Z'),
      },
    );
  });

  it('should not increment maxStreak if lastViewAt was yesterday and current streak is smaller', async () => {
    await runTest(
      '2024-01-26T19:17Z',
      '2024-01-25T17:23Z',
      {
        currentStreak: 4,
        totalStreak: 98,
        maxStreak: 10,
      },
      {
        currentStreak: 5,
        totalStreak: 99,
        maxStreak: 10,
        lastViewAt: new Date('2024-01-26T19:17Z'),
      },
    );
  });

  it('should not increment a reading streak if lastViewAt is the same day', async () => {
    await runTest('2024-01-26T19:17Z', '2024-01-26T17:23Z', defaultStreak, {
      ...defaultStreak,
      lastViewAt: new Date('2024-01-26T19:17Z'),
    });
  });

  describe('showMilestone is set correctly', () => {
    beforeEach(async () => {
      await con.getRepository(Alerts).clear();
    });

    it('should not set showStreakMilestone if current streak is same as max streak and not a fibonacci number', async () => {
      await runTest(
        '2024-01-26T19:17Z',
        '2024-01-25T17:23Z',
        { ...defaultStreak, currentStreak: 5, maxStreak: 5 },
        {
          currentStreak: 6,
          totalStreak: 43,
          maxStreak: 6,
          lastViewAt: new Date('2024-01-26T19:17Z'),
        },
      );

      const alerts = await con
        .getRepository(Alerts)
        .findOne({ where: { userId: 'u1' } });
      expect(alerts?.showStreakMilestone).toBe(false);
    });

    it('should set showStreakMilestone to true if current streak is a fibonacci number', async () => {
      await runTest('2024-01-26T19:17Z', '2024-01-25T17:23Z', defaultStreak, {
        currentStreak: 5,
        totalStreak: 43,
        maxStreak: 10,
        lastViewAt: new Date('2024-01-26T19:17Z'),
      });

      const alerts = await con
        .getRepository(Alerts)
        .findOne({ where: { userId: 'u1' } });
      expect(alerts?.showStreakMilestone).toBe(true);
    });

    it('should set showStreakMilestone to false if current streak is NOT a fibonacci number and maxStreak is bigger', async () => {
      await runTest(
        '2024-01-26T19:17Z',
        '2024-01-25T17:23Z',
        {
          ...defaultStreak,
          currentStreak: 5,
        },
        {
          currentStreak: 6,
          totalStreak: 43,
          maxStreak: 10,
          lastViewAt: new Date('2024-01-26T19:17Z'),
        },
      );

      const alerts = await con
        .getRepository(Alerts)
        .findOne({ where: { userId: 'u1' } });
      expect(alerts?.showStreakMilestone).toBe(false);
    });

    it('should set showStreakMilestone to false even if previous value is true', async () => {
      await con.getRepository(Alerts).save({
        userId: 'u1',
        showStreakMilestone: true,
      });

      await runTest(
        '2024-01-26T19:17Z',
        '2024-01-25T17:23Z',
        {
          ...defaultStreak,
          currentStreak: 5,
        },
        {
          currentStreak: 6,
          totalStreak: 43,
          maxStreak: 10,
          lastViewAt: new Date('2024-01-26T19:17Z'),
        },
      );

      const alerts = await con
        .getRepository(Alerts)
        .findOne({ where: { userId: 'u1' } });
      expect(alerts?.showStreakMilestone).toBe(false);
    });

    it('should not set showStreakMilestone if lastViewAt is the same day', async () => {
      await runTest(
        '2024-01-26T19:17Z',
        '2024-01-26T17:23Z',
        {
          ...defaultStreak,
          currentStreak: 5,
        },
        {
          ...defaultStreak,
          currentStreak: 5,
          lastViewAt: new Date('2024-01-26T19:17Z'),
        },
      );

      const alerts = await con
        .getRepository(Alerts)
        .findOne({ where: { userId: 'u1' } });
      expect(alerts).toBeNull();
    });
  });
});
