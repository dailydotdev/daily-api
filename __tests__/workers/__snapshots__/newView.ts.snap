// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reading streaks updates reading streak without a timestamp 1`] = `
Object {
  "currentStreak": 5,
  "lastViewAt": Any<Date>,
  "maxStreak": 10,
  "totalStreak": 43,
  "updatedAt": Any<Date>,
  "userId": "u1",
}
`;

exports[`should not save a new view within a week since the last view 1`] = `
View {
  "hidden": false,
  "postId": "p1",
  "referer": "referer",
  "timestamp": 2020-06-11T01:17:00.000Z,
  "userId": "u1",
}
`;

exports[`should save a new view after a week since the last view 1`] = `
View {
  "hidden": false,
  "postId": "p1",
  "referer": "referer",
  "timestamp": 2020-06-19T01:17:00.000Z,
  "userId": "u1",
}
`;

exports[`should save a new view with the provided timestamp 1`] = `
View {
  "hidden": false,
  "postId": "p1",
  "referer": "referer",
  "timestamp": 2020-06-11T01:17:00.000Z,
  "userId": "u1",
}
`;

exports[`should save a new view with the provided timestamp 2`] = `
Object {
  "currentStreak": 1,
  "lastViewAt": 2020-06-11T01:17:00.000Z,
  "maxStreak": 1,
  "totalStreak": 1,
  "updatedAt": Any<Date>,
  "userId": "u1",
}
`;

exports[`should save a new view without timestamp 1`] = `
Object {
  "hidden": false,
  "postId": "p1",
  "referer": "referer",
  "timestamp": Any<Date>,
  "userId": "u1",
}
`;
